R/add_labels.R

```diff
@@ -0,0 +1,157 @@
+#' # labelling by providing required number of labels
+#' add_labels(
+#'   x,
+#'   variable = "My x",
+#'   values = c("one", "two", "three")
+#' )
+#'
+#' # labelling using named vectors
+#' data(iris)
+#' out <- add_labels(
+#'   iris$Species,
+#'   variable = "Labelled Species",
+#'   values = c(`setosa` = "Spec1", `versicolor` = "Spec2", `virginica` = "Spec3")
+#' )
+#' str(out)
+#'
+#' # Partial labelling
+#' x <- 1:5
+#' add_labels(
+#'   x,
+#'   variable = "My x",
+#'   values = c(`1` = "lowest", `5` = "highest"),
+#'   verbose = FALSE
+#' )
+#' @export
+add_labels <- function(x, ...) {
+  UseMethod("add_labels")
+}
+
+
+#' @export
+add_labels.default <- function(x, verbose = TRUE, ...) {
+  if (isTRUE(verbose)) {
+    insight::format_alert(
+      sprintf("Adding labels currently not possible for variables of class `%s`.", class(x)[1])
+    )
+  }
+  x
+}
+
+#' @export
+add_labels.numeric <- function(x, variable = NULL, values = NULL, verbose = TRUE, ...) {
+  # add variable label
+  if (!is.null(variable)) {
+    if (is.character(variable) && length(variable) == 1) {
+      attr(x, "label") <- variable
+    } else if (verbose) {
+      insight::format_alert("Variable labels must be a character string.")
+    }
+  }
+
+  # extract unique values
+  unique_values <- as.vector(sort(stats::na.omit(unique(x))))
+  labels <- NULL
+
+  # do we have a names vector for "values"?
+  # else check if number of labels and values match
+  if (is.null(names(values))) {
+    if (length(values) == length(unique_values)) {
+      labels <- stats::setNames(unique_values, values)
+    } else if (verbose) {
+      insight::format_alert(
+        "Cannot add labels. Number of unique values and number of value labels are not equal.",
+        sprintf("There are %i unique values and %i provided labels.", length(unique_values), length(values))
+      )
+    }
+  } else {
+    # check whether we have matches of labels and values
+    matching_labels <- names(values) %in% unique_values
+    if (!all(matching_labels) && verbose) {
+      insight::format_alert(
+        "Following labels were associated with values that don't exist:",
+        text_concatenate(paste0(values[!matching_labels], " (", names(values)[!matching_labels], ")"), enclose = "`")
+      )
+    }
+    values <- values[names(values) %in% unique_values]
+
+    if (length(values)) {
+      # we need to switch names and values
+      labels <- stats::setNames(coerce_to_numeric(names(values)), values)
+    }
+  }
+
+  attr(x, "labels") <- labels
+  x
+}
+
+#' @export
+add_labels.factor <- add_labels.numeric
+
+#' @export
+add_labels.character <- add_labels.numeric
+
+#' @rdname add_labels
+#' @export
+add_labels.data.frame <- function(x,
+                                  select = NULL,
+                                  exclude = NULL,
+                                  values = NULL,
+                                  ignore_case = FALSE,
+                                  append = FALSE,
```

Comment:
what is `append`? it's not in the docs

Reply:
Would work like in other recoding/transformation functions, but maybe is less useful here? I remove it for now.